accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,706 @@
+```cs
+var fn = new delegate*<int>[1];
+Console.WriteLine(fn is delegate*<int>[]); // true
+Console.WriteLine(fn is delegate*<bool>[]); // false
+```
+```
+ldc.i4.1
+newarr     method int32 *()
+ldnull
+cgt.un
+call       void [System.Console]System.Console::WriteLine(bool)
+ldc.i4.0
+call       void [System.Console]System.Console::WriteLine(bool)
+```
+
+The C# compiler will also consider the `false` case warning CS0184: `The given expression is never of the provided...`.
+
+This C# vs. runtime disagreement with "known values" also appears with array variance:
+```cs
+Console.WriteLine(new int[1] is uint[]); // false
+Console.WriteLine(((object)new int[1]) is uint[]); // true
+```
+
+### With modifiers + unmodified types
+The above examples only compare parameter types without calling conventions or modifiers.
+
+An unmodified type does include managed vs. unmanaged as part of its type identity:
+```cs
+object o = new delegate*<int>[1];
+Console.WriteLine(o is delegate*<int>[]); // true
+Console.WriteLine(o is delegate* unmanaged[Cdecl]<int>[]); // false
+
+var fn = new delegate*<int>[1];
+Console.WriteLine(fn is delegate*<int>[]); // true
+Console.WriteLine(fn is delegate* unmanaged[Cdecl] <int>[]); // false
+```
+
+but unmanaged calling conventions are not part of the type identity:
+```cs
+object o1 = new delegate* unmanaged[SuppressGCTransition]<int>[1];
+Console.WriteLine(o1 is delegate* unmanaged[MemberFunction]<int>[]); // true
+
+object o2 = new delegate* unmanaged[CDecl]<int>[1];
+Console.WriteLine(o2 is delegate* unmanaged[StdCall]<int>[]); // true
+```
+
+In addition, all non-CallConv* modifiers, such as `ref\in\out\const` are not considered part of the type identity either.
+
+### With modified types
+A modified type acts like the underlying type with the exception that it overrides `GetFunctionPointerParameters()` to return the modified parameter types. It has an identity separate than the underlying types.
+
+```cs
+public class Holder
+{
+    public unsafe delegate* unmanaged[SuppressGCTransition]<int> _field;
+}
+
+Type modifiedType = typeof(Holder).GetField("_field").GetModifiedFieldType();
+Console.WriteLine(modifiedType.IsModifiedType); // true
+Console.WriteLine(modifiedType.UnderlyingSystemType); // (unmodified function pointer)
+Console.WriteLine(modifiedType.IsFunctionPointer); // false
+Console.WriteLine(modifiedType.UnderlyingSystemType.IsFunctionPointer); // true
+
+Console.WriteLine(modifiedType.GetFunctionPointerCallingConventions().Length); // 2
+Console.WriteLine(modifiedType.GetFunctionPointerCallingConventions[0]); // CallConvCdecl
+Console.WriteLine(modifiedType.GetFunctionPointerCallingConventions[0].IsModifiedType); // true
+Console.WriteLine(modifiedType.GetFunctionPointerCallingConventions[1]); // CallConvMemberFunction
+
+Type modifiedType2 = typeof(Holder).GetField("_field").GetModifiedFieldType(); // same field...
+Console.WriteLine(modifiedType == modifiedType2); // true (same instance)
```

Comment:
```suggestion
Console.WriteLine(modifiedType == modifiedType2); // true (types are equal)
Console.WriteLine(modifiedType.Equals(modifiedType.UnderlyingSystemType)); // true (modified and underlying types are equal)
Console.WriteLine(ReferenceEquals(modifiedType, modifiedType2)); // undefined. The implementation is not required to return same instance.
```

Operator `==` is overloaded for System.Type, so `==` does not check whether it is the same object instance.

Same instance check would be `Console.WriteLine(ReferenceEquals(modifiedType, modifiedType2))`. I do not think we need to guarantee that the implementation returns same instance.

Reply:
Sure thanks. I see that `==` is overridden and calls Equals which [inspects UnderlyingSystemType](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Type.cs#L636)